### ISSUE TITLE
Autocorrecting a Style/RedundantSelfAssignmentBranch offence breaks existing code

### ISSUE THREAD

**Issue by hugovandevliert** [rubocop/rubocop#10011]
While upgrading our codebase to RuboCop 1.19.0 and running `bundle exec rubocop -a`, I noticed some failing specs. I believe the [Style/RedundantSelfAssignmentBranch](https://docs.rubocop.org/rubocop/cops_style.html#styleredundantselfassignmentbranch) cop autocorrections should be marked unsafe. See the example below.

--------

## Expected behavior

`bundle exec rubocop -a` should not break existing code.

## Actual behavior

`bundle exec rubocop -a` broke some existing code.

## Steps to reproduce the problem

Let's say we have the following (rather odd) piece of code in `test.rb`:
```ruby
class Test
  def foo
    @foo = if @foo_set
      @foo
    else
      @foo_set = true
      @foo = 'bar'
    end
  end
end

test = Test.new
puts test.foo == 'bar' # should be true
puts test.foo == 'bar' # should be true
```

Running this file will give the following output:
```
true
true
```

RuboCop will display the following message on line 4: 
```
Style/RedundantSelfAssignmentBranch: Remove the self-assignment branch.
```

Executing `bundle exec rubocop -a test.rb` will update the code to this:
```ruby
class Test
  def foo
    @foo = unless @foo_set
      @foo_set = true
      @foo = 'bar'
    end
  end
end

test = Test.new
puts test.foo == 'bar' # should be true
puts test.foo == 'bar' # should be true
```

Running this file will give a different output:
```
true
false
```

## RuboCop version

```
1.19.0 (using Parser 3.0.2.0, rubocop-ast 1.10.0, running on ruby 2.7.4 x86_64-darwin20)
  - rubocop-performance 1.11.4
  - rubocop-rails 2.11.3
  - rubocop-rspec 2.4.0
```
